### ISSUE TITLE
Issues with specaugment_v2

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn_common#199]
While trying to use the new specaugment (https://github.com/rwth-i6/returnn_common/blob/main/asr/specaugment.py#L10) I came across multiple different errors which I can not resolve, or in some cases do not even understand. I will try to give as much details as possible. 

Also some errors belong to `returnn_common`, some to `returnn` itself. I will see how I best structure this into multiple issues, referencing to this overview issue.

Here is the basic code (both for testing full specaugment, and a single mask).

```python3
from returnn_common.asr.specaugment import random_mask_v2, specaugment_v2
[... other imports omitted ...]

data_time = nn.SpatialDim("data_time", None)
data_feature = nn.FeatureDim("data_feature", 5)
data = nn.Data(
    name="data",
    dim_tags=[nn.batch_dim, data_time, data_feature],
    available_for_inference=True,
)

class DummyNet(nn.Module):
    [...]
    def __call__(self, source: nn.Tensor, spatial_dim: nn.Dim):
        return specaugment_v2(x=source, spatial_dim=spatial_dim)
        # return random_mask_v2(x=source, mask_axis=spatial_dim, broadcast_axis=source.feature_dim, min_num=1, max_num=2, max_dims=3)

def _config_get_spec_network(epoch: int) -> dict:
    nn.reset_default_root_name_ctx()
    net = DummyNet()
    data_tensor = nn.get_extern_data(data)
    spatial_dim = data.dim_tags[data.time_dim_axis]
    out = net(source=data_tensor, spatial_dim=spatial_dim)
    out.mark_as_default_output()
    config_code = nn.get_returnn_config().get_complete_py_code_str(net)
    print(config_code)  # I will also provide this as gist
    return nn.get_returnn_config().get_net_dict_raw_dict(net)

data_args = data.get_kwargs()
data_args.pop("name")
extern_data = {"data": data_args,}

config = Config({
    "task": "train", "num_epochs": 1, "start_epoch": 1,
    "get_network": _config_get_spec_network,
    "extern_data": extern_data,
    "behavior_version": 12
})
train_dataset = init_dataset(
    {"class": "DummyDataset", "input_dim": 5, "output_dim": 12, "num_seqs": 3})
engine = Engine(config)
engine.init_train_from_config(config, train_data=train_dataset)
engine.train()
```

I create a gist containing both exceptions and both printed networks (see the print statement), so once for calling `specaugment_v2` and once for calling `random_mask_v2`:

https://gist.github.com/JackTemaki/a376d171f6c8f21d18299be9523ccbb8

```
  File "/u/rossenbach/experiments/tts_asr_2021/returnn/returnn/tf/util/data.py", line 2905, in Data.sanity_check
    line: assert tag.dyn_size is not None
    locals:
      tag = <local> Dim{'dummy_net/random_mask_v2/top_k:top_k:k_dim'[]}
      tag.dyn_size = <local> None
AssertionError
```

It is also worth noting that my error message changes when I copy the network code into a RETURNN test case. This makes debugging nearly impossible. (I am sure I am using the same RETURNN version).

The test case which is taken from the printed code for calling `random_mask_v2` is here:
https://gist.github.com/JackTemaki/43803514d4620963acf8cb0fbace0439

**Comment by albertz** [comment]
Ok, a number of individual fixes on RETURNN side:

https://github.com/rwth-i6/returnn/pull/1105
https://github.com/rwth-i6/returnn/pull/1106
https://github.com/rwth-i6/returnn/pull/1107

**Comment by albertz** [comment]
I pushed the test case `test_random_mask_v2_via_get_network`.

@JackTemaki Is this all fixed now?

**Comment by JackTemaki** [comment]
> I pushed the test case `test_random_mask_v2_via_get_network`.
> 
> @JackTemaki Is this all fixed now?

No, the error for the `random_mask_v2` case is gone, but the other error still remains.

In short:
```python
  File "/u/rossenbach/experiments/tts_asr_2021/returnn/returnn/tf/layers/base.py", line 844, in LayerBase.get_base_absolute_name_scope_prefix
    line: assert not self.name_scope.endswith("/")
    locals:
      self = <local> <InternalLayer specaugment_v2/cond/'dummy_net' out_type=Data{[B?]}>
      self.name_scope = <local> '/'
      self.name_scope.endswith = <local> <built-in method endswith of str object at 0x7f1b7c4814b0>
AssertionError
```

I think the `DummyNet` is important here.

**Comment by albertz** [comment]
Note, as mentioned, I think #147 is related here.

**Comment by albertz** [comment]
Short test case for the name_scope problem:
```python
def test_specaugment_v2_in_module_call():
  class _Net(nn.Module):
    def __call__(self, x: nn.Tensor, *, axis: nn.Dim) -> nn.Tensor:
      # specaugment_v1 uses a custom eval layer with a ref to a Python function,
      # which is defined in that module.
      # Thus, extra care needs to be taken when serializing the config.
      from ..asr.specaugment import specaugment_v2
      return specaugment_v2(x, spatial_dim=axis, global_train_step_dependent=False, only_on_train=False)

  config, net_dict = dummy_config_net_dict(net=_Net(), with_axis=True)
  dummy_run_net(config)
```

**Comment by albertz** [comment]
This should be fixed via:

* https://github.com/rwth-i6/returnn/pull/1108

**Comment by albertz** [comment]
This should all be fixed now.
